ipa-core/src/protocol/ipa_prf/mod.rs

```diff
@@ -1,6 +1,224 @@
+    C::UpgradedContext<Boolean>: UpgradedContext<Boolean, Share = Replicated<Boolean>>,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = Replicated<F>>,
+    BK: WeakSharedValue + CustomArray<Element = Boolean> + Field,
+    TV: WeakSharedValue + CustomArray<Element = Boolean> + Field,
+    TS: WeakSharedValue + CustomArray<Element = Boolean> + Field,
+    SS: WeakSharedValue + CustomArray<Element = Boolean> + Field,
+    for<'a> &'a Replicated<SS>: IntoIterator<Item = Replicated<Boolean>>,
+    for<'a> &'a Replicated<TS>: IntoIterator<Item = Replicated<Boolean>>,
+    for<'a> &'a Replicated<TV>: IntoIterator<Item = Replicated<Boolean>>,
+    for<'a> &'a Replicated<BK>: IntoIterator<Item = Replicated<Boolean>>,
+    for<'a> <&'a Replicated<SS> as IntoIterator>::IntoIter: Send,
+    for<'a> <&'a Replicated<TV> as IntoIterator>::IntoIter: Send,
+    for<'a> <&'a Replicated<TS> as IntoIterator>::IntoIter: Send,
+    F: PrimeField + ExtendableField,
+    Replicated<F>: Serializable,
+{
+    // TODO (richaj): Add shuffle either before the protocol starts or, after converting match keys to elliptical curve.
+    // We might want to do it earlier as that's a cleaner code
+
+    let prfd_inputs =
+        compute_prf_for_inputs(ctx.narrow(&Step::ConvertInputRowsToPrf), input_rows).await?;
+
+    let histogram = compute_histogram_of_users_with_row_count(&prfd_inputs);
+
+    // TODO (richaj) : Call quicksort on match keys followed by timestamp before calling attribution logic
+    attribute_cap_aggregate::<C, BK, TV, TS, SS, Replicated<F>, F>(
+        ctx,
+        prfd_inputs,
+        attribution_window_seconds,
+        &histogram,
+    )
+    .await
+}
+
+async fn compute_prf_for_inputs<C, BK, TV, TS, F>(
+    ctx: C,
+    input_rows: Vec<OprfReport<BK, TV, TS>>,
+) -> Result<Vec<PrfShardedIpaInputRow<BK, TV, TS>>, Error>
+where
+    C: UpgradableContext,
+    C::UpgradedContext<Boolean>: UpgradedContext<Boolean, Share = Replicated<Boolean>>,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = Replicated<F>>,
+    BK: WeakSharedValue + CustomArray<Element = Boolean> + Field,
+    TV: WeakSharedValue + CustomArray<Element = Boolean> + Field,
+    TS: WeakSharedValue + CustomArray<Element = Boolean> + Field,
+    for<'a> &'a Replicated<TS>: IntoIterator<Item = Replicated<Boolean>>,
+    for<'a> &'a Replicated<TV>: IntoIterator<Item = Replicated<Boolean>>,
+    for<'a> &'a Replicated<BK>: IntoIterator<Item = Replicated<Boolean>>,
+    for<'a> <&'a Replicated<TV> as IntoIterator>::IntoIter: Send,
+    for<'a> <&'a Replicated<TS> as IntoIterator>::IntoIter: Send,
+    F: PrimeField + ExtendableField,
+    Replicated<F>: Serializable,
+{
+    let ctx = ctx.set_total_records(input_rows.len());
+    let convert_ctx = ctx.narrow(&Step::ConvertFp25519);
+    let eval_ctx = ctx.narrow(&Step::EvalPrf);
+
+    let prf_key = gen_prf_key(&convert_ctx);
+
+    ctx.parallel_join(input_rows.into_iter().enumerate().map(|(idx, record)| {
```

Comment:
From performance POV, these PRfs should be calculated in parallel vs sequentially  right?